File: services/terrapod/api/routers/onboarding.py
"""AI onboarding API (#824): discover existing, unmanaged cloud resources and
generate copy-pasteable ``resource`` + ``import {}`` blocks (optionally opening
an MR).

Onboarding itself has **no feature flag** — it's gated per workspace by the
``workspace:onboard`` RBAC capability. The **AI mode** (natural-language,
conversational, config-cleanup) is the only optional part, keyed on its own
independent switch ``ai_onboarding.enabled`` — never on ``ai_summary``.

Endpoints (under /api/terrapod/v1):
    GET  /onboarding                                availability probe (any authed user)
    POST /workspaces/{id}/onboarding-sessions       start a discovery session (workspace:onboard)
    GET  /workspaces/{id}/onboarding-sessions       list a workspace's sessions (workspace:onboard)
    GET  /onboarding-sessions/{id}                  one session incl. discovery surface (workspace:onboard)

Creating a session kicks off the credential-less **D1 schema** discovery on a
scheduler trigger (off the request thread); the client polls the session until
``status == "schema_ready"``, then reads the discovery surface. The surface is
served from a time-limited Redis cache (never persisted per-session); a session
whose cache entry has expired simply re-runs discovery. D2/D3 (query + generate)
run on the runner in a later phase.
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from terrapod.api.dependencies import AuthenticatedUser, get_current_user
from terrapod.api.pagination import paginate
from terrapod.api.serialization import rfc3339
from terrapod.auth.capabilities import WORKSPACE_ONBOARD, has_capability
from terrapod.config import settings
from terrapod.db.models import OnboardingSession, Workspace
from terrapod.db.session import get_db
from terrapod.logging_config import get_logger
from terrapod.services import onboarding_polish, onboarding_service
from terrapod.services.workspace_rbac_service import resolve_workspace_capabilities_for

router = APIRouter(tags=["onboarding"])
logger = get_logger(__name__)


@router.get("/onboarding")
async def onboarding_availability(
    user: AuthenticatedUser = Depends(get_current_user),
) -> dict:
    """Report onboarding availability to the UI.

    The onboarding feature is always present (no feature flag) — whether a given
    user can actually run it is decided per-workspace by the ``workspace:onboard``
    capability on the real endpoints. This probe reports whether the optional
    **AI mode** is available (its own switch + a configured model), so the UI can
    offer the conversational path when it's set up.
    """
    cfg = settings.ai_onboarding
    return {
        "data": {
            "type": "onboarding-availability",
            "attributes": {
                "ai-available": cfg.enabled,
                "ai-model-configured": bool(cfg.enabled and cfg.model),
            },
        }
    }


# --- helpers ---------------------------------------------------------------
async def _get_workspace(workspace_id: str, db: AsyncSession) -> Workspace:
    ws_uuid = workspace_id.removeprefix("ws-")
    result = await db.execute(select(Workspace).where(Workspace.id == ws_uuid))
    ws = result.scalar_one_or_none()
    if ws is None:
        raise HTTPException(status_code=404, detail="Workspace not found")
    return ws


async def _require_onboard(ws: Workspace, user: AuthenticatedUser, db: AsyncSession) -> None:
    caps = await resolve_workspace_capabilities_for(db, user, ws)
    if not has_capability(caps, WORKSPACE_ONBOARD):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"Requires '{WORKSPACE_ONBOARD}' capability on workspace",
        )


def _session_json(
    s: OnboardingSession, surface: dict | None = None, *, detail: bool = False
) -> dict:
    """Serialize a session. ``surface`` (from the Redis cache) and the derived
    ``paired-*`` views are included only on the **detail** read — the surface is
    never stored on the row, and the paired views are a regex reflow over the full
    config, so computing them per-row in the list handler (an ``async def``) would
    be sync work on the event loop (Rule 13). The list omits both."""
    paired_config = (
        onboarding_polish.pair_config_and_imports(s.generated_config, s.import_blocks or "")
        if detail and s.generated_config
        else None
    )
    paired_polished = (
        onboarding_polish.pair_config_and_imports(s.polished_config, s.polished_import_blocks or "")
        if detail and s.polished_config
        else None
    )
    return {
        "type": "onboarding-sessions",
        "id": str(s.id),
        "attributes": {
            "workspace-id": str(s.workspace_id),
            "status": s.status,
            "provider": s.provider,
            "provider-version": s.provider_version,
            "engine": s.engine,
            "engine-version": s.engine_version,
            "selected-types": s.selected_types or [],
            "ai-assisted": s.ai_assisted,
            "error": s.error,
            "data-source-count": (surface or {}).get("count") if surface else None,
            "discovery-surface": surface,
            # D3 output — the reviewable payload. Present once status == config_ready.
            "generated-config": s.generated_config,
            "import-blocks": s.import_blocks,
            # Optional AI-polished view (#824 Phase A) — resources renamed from
            # tags, grouped, commented; import ids/values untouched. Null until
            # the polish lands (or if rejected / AI disabled). `ai-assisted`
            # (above) is true iff these are populated.
            "polished-config": s.polished_config,
            "polished-import-blocks": s.polished_import_blocks,
            # Derived, presentation-only "paired" view: each `import {}` interleaved
            # directly above the resource it targets (import ids/values untouched;
            # the split fields above stay canonical). Computed at serialize time —
            # never stored. Null until config exists.
            "paired-config": paired_config,
            "paired-polished-config": paired_polished,
            "discovery-run-id": str(s.discovery_run_id) if s.discovery_run_id else None,
            "result-run-id": str(s.result_run_id) if s.result_run_id else None,
            "created-by": s.created_by,
            "created-at": rfc3339(s.created_at),
            "updated-at": rfc3339(s.updated_at),
        },
    }


# --- endpoints -------------------------------------------------------------
@router.post("/workspaces/{workspace_id}/onboarding-sessions", status_code=201)
async def create_onboarding_session(
    workspace_id: str,
    body: dict,
    db: AsyncSession = Depends(get_db),
    user: AuthenticatedUser = Depends(get_current_user),
) -> dict:
    """Start a discovery session and kick off credential-less D1 schema discovery."""
    ws = await _get_workspace(workspace_id, db)
    await _require_onboard(ws, user, db)

    attrs = (body.get("data") or {}).get("attributes") or {}
    provider = str(attrs.get("provider") or "").strip()
    if not onboarding_service.is_valid_provider(provider):
        raise HTTPException(
            status_code=422,
            detail="provider must be a lowercase terraform provider name (e.g. 'aws')",
        )
    provider_version = str(
        attrs.get("provider-version") or attrs.get("provider_version") or ""
    ).strip()
    if not onboarding_service.is_valid_version_constraint(provider_version):
        raise HTTPException(
            status_code=422,
            detail="provider-version must be a terraform version constraint (e.g. '< 6.0', '~> 5.0')",
        )

    session = await onboarding_service.create_session(
        db,
        workspace_id=ws.id,
        provider=provider,
        created_by=user.email,
        provider_version=provider_version,
    )
    await db.commit()

    # Run the (potentially slow, first-time) schema discovery off the request
    # thread. Best-effort enqueue — the session is already persisted as pending.
    try:
        from terrapod.services.scheduler import enqueue_trigger

        await enqueue_trigger(
            "onboarding_schema_discover",
            {"session_id": str(session.id)},
            dedup_key=f"onb_schema:{session.id}",
        )
    except Exception as exc:  # noqa: BLE001 — surface stays pending, retriable
        logger.warning(
            "onboarding_schema_enqueue_failed", session_id=str(session.id), error=str(exc)
        )

    return {"data": _session_json(session)}


@router.get("/workspaces/{workspace_id}/onboarding-sessions")
async def list_onboarding_sessions(
    workspace_id: str,
    request: Request = None,
    db: AsyncSession = Depends(get_db),
    user: AuthenticatedUser = Depends(get_current_user),
) -> dict:
    ws = await _get_workspace(workspace_id, db)
    await _require_onboard(ws, user, db)
    sessions = await onboarding_service.list_sessions(db, ws.id)
    # Surfaces are omitted from the list (each is a large per-session Redis read);
    # the detail endpoint carries the surface.
    items = [_session_json(s) for s in sessions]
    page_items, meta = paginate(items, request)
    return {"data": page_items, "meta": meta}


@router.get("/onboarding-sessions/{session_id}")
async def get_onboarding_session(
    session_id: str,
    db: AsyncSession = Depends(get_db),
    user: AuthenticatedUser = Depends(get_current_user),
) -> dict:
    try:
        sid = uuid.UUID(session_id)
    except ValueError:
        raise HTTPException(status_code=404, detail="Onboarding session not found") from None
    session = await onboarding_service.get_session(db, sid)
    if session is None:
        raise HTTPException(status_code=404, detail="Onboarding session not found")
    ws = await db.get(Workspace, session.workspace_id)
    if ws is None:
        raise HTTPException(status_code=404, detail="Workspace not found")
    await _require_onboard(ws, user, db)
    surface = await onboarding_service.get_session_surface(session)
    return {"data": _session_json(session, surface=surface, detail=True)}


@router.post("/onboarding-sessions/{session_id}/discover")
async def start_onboarding_discovery(
    session_id: str,
    body: dict,
    db: AsyncSession = Depends(get_db),
    user: AuthenticatedUser = Depends(get_current_user),
) -> dict:
    """Dispatch the D2/D3 runner discovery run for a schema_ready session.

    Body: ``{"data": {"attributes": {"selected-types": ["aws_vpcs", ...]}}}`` —
    the data-source types (a subset of the session's discovery surface) to query.
    """
    try:
        sid = uuid.UUID(session_id)
    except ValueError:
        raise HTTPException(status_code=404, detail="Onboarding session not found") from None
    session = await onboarding_service.get_session(db, sid)
    if session is None:
        raise HTTPException(status_code=404, detail="Onboarding session not found")
    ws = await db.get(Workspace, session.workspace_id)
    if ws is None:
        raise HTTPException(status_code=404, detail="Workspace not found")
    await _require_onboard(ws, user, db)

    attrs = (body.get("data") or {}).get("attributes") or {}
    raw_types = attrs.get("selected-types") or attrs.get("selected_types") or []
    if not isinstance(raw_types, list) or not all(isinstance(t, str) for t in raw_types):
        raise HTTPException(status_code=422, detail="selected-types must be a list of strings")

    try:
        session = await onboarding_service.start_discovery(db, session, raw_types)
    except onboarding_service.OnboardingError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from None
    await db.commit()

    surface = await onboarding_service.get_session_surface(session)
    return {"data": _session_json(session, surface=surface, detail=True)}

File: go-terrapod/onboarding.go
package terrapod

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
)

// OnboardingAvailability reports whether the optional AI onboarding mode is
// available. Onboarding itself has no feature flag — it is gated per workspace
// by the workspace:onboard capability — so this probe only describes the AI
// path (its own switch + a configured model).
type OnboardingAvailability struct {
	AIAvailable       bool `json:"ai-available"`
	AIModelConfigured bool `json:"ai-model-configured"`
}

// OnboardingDataSource is one entry in a session's discovery surface: a data
// source the deterministic import path can consume (it exposes a derivable
// resource-id list and singularises to a real managed resource). The field tags
// mirror the snake_case JSON emitted by `terrapod-query schema --importable` and
// cached server-side, NOT the kebab-case JSON:API attribute convention used
// elsewhere in this SDK.
type OnboardingDataSource struct {
	Provider       string   `json:"provider"`
	Name           string   `json:"name"`
	HasFilter      bool     `json:"has_filter"`
	HasTags        bool     `json:"has_tags"`
	ReturnsList    bool     `json:"returns_list"`
	IDListAttr     string   `json:"id_list_attr,omitempty"`
	ResourceType   string   `json:"resource_type,omitempty"`
	Inputs         []string `json:"inputs,omitempty"`
	RequiredInputs []string `json:"required_inputs,omitempty"`
}

// DiscoverySurface is the credential-less D1 schema-discovery result for a
// session: the importable data sources the caller may query. Present on the
// detail read (GetOnboardingSession / StartOnboardingDiscovery) once the session
// reaches schema_ready; nil on the list read (the surface is a large per-session
// value the list endpoint omits).
type DiscoverySurface struct {
	Count       int                    `json:"count"`
	DataSources []OnboardingDataSource `json:"data_sources"`
}

// OnboardingSession is the decoded form of one onboarding / resource-discovery
// session. Field tags mirror the JSON:API attribute names.
//
// Nullable value fields use pointers where nil is meaningful: DataSourceCount
// and DiscoverySurface are nil until D1 schema discovery completes (and stay nil
// on the list read, which omits the surface). Config fields are empty strings
// until the D2/D3 discovery run produces them.
type OnboardingSession struct {
	ID              string   `json:"id"`
	WorkspaceID     string   `json:"workspace-id"`
	Status          string   `json:"status"`
	Provider        string   `json:"provider"`
	ProviderVersion string   `json:"provider-version,omitempty"`
	Engine          string   `json:"engine,omitempty"`
	EngineVersion   string   `json:"engine-version,omitempty"`
	SelectedTypes   []string `json:"selected-types,omitempty"`
	AIAssisted      bool     `json:"ai-assisted"`
	Error           string   `json:"error,omitempty"`
	// DataSourceCount is the number of importable data sources the surface
	// found; nil until schema discovery completes (and on the list read).
	DataSourceCount *int64 `json:"data-source-count,omitempty"`
	// DiscoverySurface is the full importable-data-source catalogue; nil on the
	// list read and until schema_ready (see DiscoverySurface).
	DiscoverySurface *DiscoverySurface `json:"discovery-surface,omitempty"`
	// GeneratedConfig and ImportBlocks are the reviewable D3 output — the
	// resource config and the matching import {} blocks. Empty until the
	// session reaches config_ready.
	GeneratedConfig string `json:"generated-config,omitempty"`
	ImportBlocks    string `json:"import-blocks,omitempty"`
	// PolishedConfig / PolishedImportBlocks are the optional AI-polished view
	// (resources renamed from tags, grouped, commented; import ids untouched).
	// Empty until the polish lands (or if rejected / AI disabled). AIAssisted is
	// true iff these are populated.
	PolishedConfig       string `json:"polished-config,omitempty"`
	PolishedImportBlocks string `json:"polished-import-blocks,omitempty"`
	// PairedConfig / PairedPolishedConfig are the presentation-only view with
	// each import {} block interleaved above the resource it targets. Computed
	// server-side at serialize time; empty until config exists.
	PairedConfig         string `json:"paired-config,omitempty"`
	PairedPolishedConfig string `json:"paired-polished-config,omitempty"`
	// DiscoveryRunID is the D2/D3 discovery run; ResultRunID is the eventual
	// import-only run. Empty until each is dispatched.
	DiscoveryRunID string `json:"discovery-run-id,omitempty"`
	ResultRunID    string `json:"result-run-id,omitempty"`
	CreatedBy      string `json:"created-by,omitempty"`
	CreatedAt      string `json:"created-at,omitempty"`
	UpdatedAt      string `json:"updated-at,omitempty"`
}

// CreateOnboardingSessionRequest starts a discovery session for a workspace and
// kicks off credential-less D1 schema discovery. Provider is a lowercase
// terraform provider name (e.g. "aws"); ProviderVersion is an optional terraform
// version constraint (e.g. "~> 5.0", "< 6.0"). Both are validated server-side
// (422 on bad input).
type CreateOnboardingSessionRequest struct {
	WorkspaceID     string
	Provider        string
	ProviderVersion string
}

// GetOnboardingAvailability reports whether the optional AI onboarding mode is
// available (any authenticated user).
func (c *Client) GetOnboardingAvailability(ctx context.Context) (*OnboardingAvailability, error) {
	data, err := c.Get(ctx, "/api/terrapod/v1/onboarding")
	if err != nil {
		return nil, err
	}
	res, err := ParseResource(data)
	if err != nil {
		return nil, fmt.Errorf("parse onboarding availability response: %w", err)
	}
	return &OnboardingAvailability{
		AIAvailable:       GetBoolAttr(res, "ai-available"),
		AIModelConfigured: GetBoolAttr(res, "ai-model-configured"),
	}, nil
}

// CreateOnboardingSession starts a discovery session and kicks off D1 schema
// discovery off the request thread. Poll GetOnboardingSession until Status ==
// "schema_ready", then read DiscoverySurface. Requires the workspace:onboard
// capability on the workspace.
func (c *Client) CreateOnboardingSession(ctx context.Context, req CreateOnboardingSessionRequest) (*OnboardingSession, error) {
	if req.WorkspaceID == "" {
		return nil, fmt.Errorf("workspace id is required")
	}
	attrs := map[string]any{"provider": req.Provider}
	if req.ProviderVersion != "" {
		attrs["provider-version"] = req.ProviderVersion
	}
	body, err := MarshalResource("onboarding-sessions", attrs, nil)
	if err != nil {
		return nil, fmt.Errorf("marshal create onboarding-session: %w", err)
	}
	data, err := c.Post(ctx,
		fmt.Sprintf("/api/terrapod/v1/workspaces/%s/onboarding-sessions", url.PathEscape(req.WorkspaceID)),
		body)
	if err != nil {
		return nil, err
	}
	return parseOnboardingSession(data)
}

// ListOnboardingSessions returns a workspace's discovery sessions, newest-ish
// first (server order). The discovery surface is omitted from list entries; read
// a single session for it. Requires workspace:onboard.
func (c *Client) ListOnboardingSessions(ctx context.Context, workspaceID string) ([]OnboardingSession, error) {
	if workspaceID == "" {
		return nil, fmt.Errorf("workspace id is required")
	}
	data, err := c.Get(ctx,
		fmt.Sprintf("/api/terrapod/v1/workspaces/%s/onboarding-sessions", url.PathEscape(workspaceID)))
	if err != nil {
		return nil, err
	}
	resources, err := ParseResourceList(data)
	if err != nil {
		return nil, fmt.Errorf("parse onboarding-sessions list: %w", err)
	}
	out := make([]OnboardingSession, 0, len(resources))
	for i := range resources {
		out = append(out, *onboardingSessionFromResource(&resources[i]))
	}
	return out, nil
}

// GetOnboardingSession reads one session by id, including the discovery surface
// (from the time-limited server cache; an expired cache simply re-runs
// discovery). Requires workspace:onboard on the session's workspace.
func (c *Client) GetOnboardingSession(ctx context.Context, sessionID string) (*OnboardingSession, error) {
	if sessionID == "" {
		return nil, fmt.Errorf("session id is required")
	}
	data, err := c.Get(ctx, "/api/terrapod/v1/onboarding-sessions/"+url.PathEscape(sessionID))
	if err != nil {
		return nil, err
	}
	return parseOnboardingSession(data)
}

// StartOnboardingDiscovery dispatches the D2/D3 runner discovery run for a
// schema_ready session. selectedTypes is the subset of the session's discovery
// surface (data-source types) to query — it must be a non-empty subset, else the
// server returns 422. Requires workspace:onboard.
func (c *Client) StartOnboardingDiscovery(ctx context.Context, sessionID string, selectedTypes []string) (*OnboardingSession, error) {
	if sessionID == "" {
		return nil, fmt.Errorf("session id is required")
	}
	body, err := MarshalResource("onboarding-sessions", map[string]any{
		"selected-types": selectedTypes,
	}, nil)
	if err != nil {
		return nil, fmt.Errorf("marshal start onboarding-discovery: %w", err)
	}
	data, err := c.Post(ctx,
		fmt.Sprintf("/api/terrapod/v1/onboarding-sessions/%s/discover", url.PathEscape(sessionID)),
		body)
	if err != nil {
		return nil, err
	}
	return parseOnboardingSession(data)
}

// ── Internal helpers ─────────────────────────────────────────────────

func parseOnboardingSession(body []byte) (*OnboardingSession, error) {
	res, err := ParseResource(body)
	if err != nil {
		return nil, fmt.Errorf("parse onboarding-session response: %w", err)
	}
	return onboardingSessionFromResource(res), nil
}

func onboardingSessionFromResource(res *Resource) *OnboardingSession {
	s := &OnboardingSession{
		ID:                   res.ID,
		WorkspaceID:          GetStringAttr(res, "workspace-id"),
		Status:               GetStringAttr(res, "status"),
		Provider:             GetStringAttr(res, "provider"),
		ProviderVersion:      GetStringAttr(res, "provider-version"),
		Engine:               GetStringAttr(res, "engine"),
		EngineVersion:        GetStringAttr(res, "engine-version"),
		SelectedTypes:        GetListAttr(res, "selected-types"),
		AIAssisted:           GetBoolAttr(res, "ai-assisted"),
		Error:                GetStringAttr(res, "error"),
		GeneratedConfig:      GetStringAttr(res, "generated-config"),
		ImportBlocks:         GetStringAttr(res, "import-blocks"),
		PolishedConfig:       GetStringAttr(res, "polished-config"),
		PolishedImportBlocks: GetStringAttr(res, "polished-import-blocks"),
		PairedConfig:         GetStringAttr(res, "paired-config"),
		PairedPolishedConfig: GetStringAttr(res, "paired-polished-config"),
		DiscoveryRunID:       GetStringAttr(res, "discovery-run-id"),
		ResultRunID:          GetStringAttr(res, "result-run-id"),
		CreatedBy:            GetStringAttr(res, "created-by"),
		CreatedAt:            GetStringAttr(res, "created-at"),
		UpdatedAt:            GetStringAttr(res, "updated-at"),
	}
	// Nullable count: present (non-null) only once schema discovery completes.
	if raw, ok := res.Attributes["data-source-count"]; ok && len(raw) > 0 && string(raw) != "null" {
		var n int64
		if json.Unmarshal(raw, &n) == nil {
			s.DataSourceCount = &n
		}
	}
	// Nested surface object: present only on the detail read once schema_ready.
	if raw, ok := res.Attributes["discovery-surface"]; ok && len(raw) > 0 && string(raw) != "null" {
		var surf DiscoverySurface
		if json.Unmarshal(raw, &surf) == nil {
			s.DiscoverySurface = &surf
		}
	}
	return s
}
